<a href="https://colab.research.google.com/github/ZainAli24/OpenAI_Agents_SDK_class_02/blob/main/Use_LiteLLM_with_OpenAI_Agent_SDK%2C_Basic_Agent%2C_Agent_with_Tool%2C_Agent_Loop%2C_Handoffs_Class_12.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. !pip install -Uq openai-agents "openai-agents[litellm]" breakdown:**

Chaliye, is command ko break down karke concise tareeke se samjhaata hoon:

```
!pip install -Uq openai-agents "openai-agents[litellm]"
```

### **Breakdown**
Yeh ek pip command hai jo Python packages install karta hai. Isko parts mein dekhte hain:

1. **`!`**:
   - Yeh Jupyter Notebook ya IPython environment mein use hota hai taake shell command (jaise `pip`) notebook ke andar chala sakein.
   - Agar aap normal terminal mein hain, to `!` nahi likhenge.

2. **`pip install`**:
   - Yeh Python ke package manager `pip` ka command hai jo packages ko install karta hai.

3. **`-U`**:
   - Yeh flag ka matlab hai **"upgrade"**. Agar `openai-agents` package pehle se install hai, to yeh usko latest version mein update kar dega.

4. **`-q`**:
   - Yeh flag ka matlab hai **"quiet"**. Iska kaam hai installation ke output ko kam karna, taake terminal mein zyada logs na dikhein. Yani, chup-chaap install ho jayega.

5. **`openai-agents`**:
   - Yeh pehla package hai jo install ho raha hai. `openai-agents` ek SDK (software development kit) hai jo OpenAI ke tools ya agents ke saath kaam karne ke liye banaya gaya hai.

6. **`"openai-agents[litellm]"`**:
   - Yeh doosra package ya configuration hai jo install ho raha hai.
   - `[litellm]` ka matlab hai ke `openai-agents` ke saath ek optional dependency ya extra feature install ho raha hai, jo `litellm` naam ke package se related hai.
   - `litellm` ek lightweight library ho sakti hai jo `openai-agents` ke saath kaam karti hai, maslan, model integration ya API calls ke liye.

### **Concise Samajh**
- **Pehli Baat**: `!pip install -Uq openai-agents` ne `openai-agents` SDK ko install ya update kiya, quietly (bina zyada output ke).
- **Doosri Baat**: `"openai-agents[litellm]"` ne `openai-agents` ke saath `litellm` naam ka extra feature ya dependency bhi install kiya, jo is SDK ke kuch specific functionalities ke liye zaroori hai.

Bas itna hi! Yeh command do packages ko ek saath set up karta hai taake aap `openai-agents` aur uske `litellm` features ka use kar sakein.

# **1. Install openai-agents SDK with Litellm:**

In [1]:
!pip install -Uq openai-agents  "openai-agents[litellm]"


# **2. Make your Jupyter Notebook capable of running asynchronous functions:**

In [2]:
import nest_asyncio
nest_asyncio.apply()

# **3. Run Google Gemini with LiteLLm and OPENAI-Agent SDK:**

## **Function Run Sync:**

In [3]:
import asyncio
from agents import Agent, Runner, set_tracing_disabled
from agents.extensions.models.litellm_model import LitellmModel
from google.colab import userdata

MODEL = 'gemini/gemini-2.0-flash'
GEMINI_API_KEY = userdata.get("GEMINI_API_KEY")

set_tracing_disabled(disabled=True)


def starter(model, api_key):

  agent = Agent(
      name="Helper",
      instructions="You are Helpful Assistant in AI Cars Researcher",
      model=LitellmModel(model=model, api_key=api_key)
  )

  result= Runner.run_sync(    # Run AgentLoop
      agent,
      "Which Company acheive fully make autonomus car"
  )
  print(result.final_output)


starter(model=MODEL, api_key=GEMINI_API_KEY)

As of my last knowledge update in early 2023, no company has *fully* achieved a completely autonomous car, meaning a Level 5 autonomous vehicle that can drive anywhere, anytime, in any conditions, without any human intervention.

However, several companies are making significant progress and have deployed vehicles with advanced autonomous capabilities:

*   **Waymo (Alphabet):** Waymo is widely considered a leader in autonomous driving technology. They operate a ride-hailing service called Waymo One in limited areas where their vehicles can drive without a safety driver in the car. However, these areas are geofenced and well-mapped, and the vehicles still encounter limitations in certain conditions.

*   **Tesla:** Tesla has deployed its "Full Self-Driving" (FSD) Beta to a limited number of users. While Tesla calls it FSD, it is not truly fully autonomous. It is a Level 2 system requiring active driver supervision, and it has faced scrutiny and regulatory investigations due to safety c

# **Use tool and Run Async:**



### **Is mien hum ne do ceezay sekhi hai:**
----------
### **1. Tools yani function calling kese integret kar sakhte hai**
----------
### **2. Seond pe hume ne dekha ke jo AgentLoop hai woh baar baar LLM ko call kar rha hai aur reasoning engine ke taur par Agent loop LLM ko use kar rha hai:**
### is mien jab hum ne Runner se agent loop ko run kia toh agent query le ke aur jo us ko tool provide kia tha wo leke LLM ke pass gia , LLM ne query ko dekhte howe agent se bola ke ke ye tool call karke lao, aur phir agent ne wo tool call kia aur us tool ka answer wapis LLM ke pass le aya aur pehle ki sari conversation bhi agent LLM ke pass ke aya , phir LLM ne tool ke answer aur pehle ki conversation dekhte howe better way mein user ki query ka answer dia hai.

--------

In [14]:
from agents import Runner, Agent, function_tool, set_tracing_disabled
from agents.extensions.models.litellm_model import LitellmModel
from google.colab import userdata
import asyncio

from agents import enable_verbose_stdout_logging
enable_verbose_stdout_logging()

set_tracing_disabled(disabled=True)

MODEL = 'gemini/gemini-2.0-flash'
GEMINI_API_KEY = userdata.get("GEMINI_API_KEY")

@function_tool
def wheather(city: str):
  print(f"\n Getting wheather for {city} \n\n")
  return f"The wheather in {city} is sunny and warm"


async def tooler(model, api_key):
  agent = Agent(
      name="Assistant",
      instructions="You only respond in haikus.",
      model= LitellmModel(model=model, api_key=api_key),
      tools=[wheather],
  )

  result = await Runner.run(
      agent,
      "What`s the wheather of Karachi"
  )

  print(result.final_output)


asyncio.run(tooler(model=MODEL, api_key=GEMINI_API_KEY))

Tracing is disabled. Not creating trace Agent workflow


DEBUG:openai.agents:Tracing is disabled. Not creating trace Agent workflow


Setting current trace: no-op


DEBUG:openai.agents:Setting current trace: no-op


Tracing is disabled. Not creating span <agents.tracing.span_data.AgentSpanData object at 0x78130c2f8ad0>


DEBUG:openai.agents:Tracing is disabled. Not creating span <agents.tracing.span_data.AgentSpanData object at 0x78130c2f8ad0>


Running agent Assistant (turn 1)


DEBUG:openai.agents:Running agent Assistant (turn 1)


Tracing is disabled. Not creating span <agents.tracing.span_data.GenerationSpanData object at 0x78130c2f8e90>


DEBUG:openai.agents:Tracing is disabled. Not creating span <agents.tracing.span_data.GenerationSpanData object at 0x78130c2f8e90>


Calling Litellm model: gemini/gemini-2.0-flash
[
  {
    "content": "You only respond in haikus.",
    "role": "system"
  },
  {
    "role": "user",
    "content": "What`s the wheather of Karachi"
  }
]
Tools:
[
  {
    "type": "function",
    "function": {
      "name": "wheather",
      "description": "",
      "parameters": {
        "properties": {
          "city": {
            "title": "City",
            "type": "string"
          }
        },
        "required": [
          "city"
        ],
        "title": "wheather_args",
        "type": "object",
        "additionalProperties": false
      }
    }
  }
]
Stream: False
Tool choice: NOT_GIVEN
Response format: NOT_GIVEN



DEBUG:openai.agents:Calling Litellm model: gemini/gemini-2.0-flash
[
  {
    "content": "You only respond in haikus.",
    "role": "system"
  },
  {
    "role": "user",
    "content": "What`s the wheather of Karachi"
  }
]
Tools:
[
  {
    "type": "function",
    "function": {
      "name": "wheather",
      "description": "",
      "parameters": {
        "properties": {
          "city": {
            "title": "City",
            "type": "string"
          }
        },
        "required": [
          "city"
        ],
        "title": "wheather_args",
        "type": "object",
        "additionalProperties": false
      }
    }
  }
]
Stream: False
Tool choice: NOT_GIVEN
Response format: NOT_GIVEN



LLM resp:
{
  "content": null,
  "role": "assistant",
  "tool_calls": [
    {
      "index": 0,
      "function": {
        "arguments": "{\"city\": \"Karachi\"}",
        "name": "wheather"
      },
      "id": "call_68170b4a-365c-493d-8985-97a7a5322d2a",
      "type": "function"
    }
  ],
  "function_call": null
}



DEBUG:openai.agents:LLM resp:
{
  "content": null,
  "role": "assistant",
  "tool_calls": [
    {
      "index": 0,
      "function": {
        "arguments": "{\"city\": \"Karachi\"}",
        "name": "wheather"
      },
      "id": "call_68170b4a-365c-493d-8985-97a7a5322d2a",
      "type": "function"
    }
  ],
  "function_call": null
}



Tracing is disabled. Not creating span <agents.tracing.span_data.FunctionSpanData object at 0x78130c2f9e50>


DEBUG:openai.agents:Tracing is disabled. Not creating span <agents.tracing.span_data.FunctionSpanData object at 0x78130c2f9e50>


Invoking tool wheather with input {"city": "Karachi"}


DEBUG:openai.agents:Invoking tool wheather with input {"city": "Karachi"}


Tool call args: ['Karachi'], kwargs: {}


DEBUG:openai.agents:Tool call args: ['Karachi'], kwargs: {}



 Getting wheather for Karachi 


Tool wheather returned The wheather in Karachi is sunny and warm


DEBUG:openai.agents:Tool wheather returned The wheather in Karachi is sunny and warm


Running agent Assistant (turn 2)


DEBUG:openai.agents:Running agent Assistant (turn 2)


Tracing is disabled. Not creating span <agents.tracing.span_data.GenerationSpanData object at 0x78130c2f8b90>


DEBUG:openai.agents:Tracing is disabled. Not creating span <agents.tracing.span_data.GenerationSpanData object at 0x78130c2f8b90>


Calling Litellm model: gemini/gemini-2.0-flash
[
  {
    "content": "You only respond in haikus.",
    "role": "system"
  },
  {
    "role": "user",
    "content": "What`s the wheather of Karachi"
  },
  {
    "role": "assistant",
    "tool_calls": [
      {
        "id": "call_68170b4a-365c-493d-8985-97a7a5322d2a",
        "type": "function",
        "function": {
          "name": "wheather",
          "arguments": "{\"city\": \"Karachi\"}"
        }
      }
    ]
  },
  {
    "role": "tool",
    "tool_call_id": "call_68170b4a-365c-493d-8985-97a7a5322d2a",
    "content": "The wheather in Karachi is sunny and warm"
  }
]
Tools:
[
  {
    "type": "function",
    "function": {
      "name": "wheather",
      "description": "",
      "parameters": {
        "properties": {
          "city": {
            "title": "City",
            "type": "string"
          }
        },
        "required": [
          "city"
        ],
        "title": "wheather_args",
        "type": "object"
      }


DEBUG:openai.agents:Calling Litellm model: gemini/gemini-2.0-flash
[
  {
    "content": "You only respond in haikus.",
    "role": "system"
  },
  {
    "role": "user",
    "content": "What`s the wheather of Karachi"
  },
  {
    "role": "assistant",
    "tool_calls": [
      {
        "id": "call_68170b4a-365c-493d-8985-97a7a5322d2a",
        "type": "function",
        "function": {
          "name": "wheather",
          "arguments": "{\"city\": \"Karachi\"}"
        }
      }
    ]
  },
  {
    "role": "tool",
    "tool_call_id": "call_68170b4a-365c-493d-8985-97a7a5322d2a",
    "content": "The wheather in Karachi is sunny and warm"
  }
]
Tools:
[
  {
    "type": "function",
    "function": {
      "name": "wheather",
      "description": "",
      "parameters": {
        "properties": {
          "city": {
            "title": "City",
            "type": "string"
          }
        },
        "required": [
          "city"
        ],
        "title": "wheather_args",
        "type

LLM resp:
{
  "content": "Karachi is warm now,\nThe sun shines bright in the sky,\nA beautiful day.\n",
  "role": "assistant",
  "tool_calls": null,
  "function_call": null
}



DEBUG:openai.agents:LLM resp:
{
  "content": "Karachi is warm now,\nThe sun shines bright in the sky,\nA beautiful day.\n",
  "role": "assistant",
  "tool_calls": null,
  "function_call": null
}



Resetting current trace


DEBUG:openai.agents:Resetting current trace


Karachi is warm now,
The sun shines bright in the sky,
A beautiful day.



# **Handoffs:**
## is mien hamare pass ik **manager Agent** hota hai jo ke decide karta hai task ki base par ke ye task kis agent ko assign karna hai, jo agent is task ko karne ke capable hoga us ko Manager Agent ye task delegate karde ga.

In [5]:
from agents import Runner, Agent, function_tool, set_tracing_disabled
from agents.extensions.models.litellm_model import LitellmModel
from google.colab import userdata
import asyncio

set_tracing_disabled(disabled=True)

MODEL = 'gemini/gemini-2.0-flash'
GEMINI_API_KEY = userdata.get("GEMINI_API_KEY")

@function_tool()
def weather(city: str):
    print(f"\n Getting weather for {city} \n\n")
    return f"The weather in {city} is sunny and warm"

@function_tool()
def calculate_square(val: int):
    print("\n Calculating square \n\n")
    return f"The square of {val} is {val * val}"

async def tooler(model, api_key):
    WeatherAssistant = Agent(
        name="WeatherAssistant",
        instructions="You are a powerful weather assistant. Use the weather tool to provide accurate weather information for the location specified in the user's query.",
        model=LitellmModel(model=model, api_key=api_key),
        tools=[weather],
        handoff_description="Provides current weather conditions, forecasts, and climate information for specific locations."
    )

    Calculator = Agent(
        name="Calculator",
        instructions="You are a powerful math assistant. Use the calculate_square tool to compute the square of the number provided in the user's query.",
        model=LitellmModel(model=model, api_key=api_key),
        tools=[calculate_square],
        handoff_description="Performs mathematical calculations, specifically computing the square of numbers."
    )

    ManagerAgent = Agent(
        name="Manager",
        instructions="You are a manager agent. Analyze the user's query and decide whether to handle it yourself or delegate to a specialized agent. If the query is about weather, use the transfer_to_WeatherAssistant tool. If it involves calculating the square of a number, use the transfer_to_Calculator tool. For other general questions, you can answer directly.",
        model=LitellmModel(model=model, api_key=api_key),
        handoffs=[WeatherAssistant, Calculator],
    )

    result = await Runner.run(
        ManagerAgent,
        "What`s the square of 24"
    )

    print(result.final_output)

asyncio.run(tooler(model=MODEL, api_key=GEMINI_API_KEY))


 Calculating square 


The square of 24 is 576.



In [13]:
@function_tool()
def weather(city: str):
    print(f"\n Getting weather for {city} \n\n")
    return f"The weather in {city} is sunny and warm"

@function_tool()
def calculate_square(val: int):
    print("\n Calculating square \n\n")
    return f"The square of {val} is {val * val}"


WeatherAssistant = Agent(
    name="WeatherAssistant",
    instructions="You are a powerful weather assistant. Use the weather tool to provide accurate weather information for the location specified in the user's query.",
    model=LitellmModel(model=MODEL, api_key=GEMINI_API_KEY),
    tools=[weather],
    handoff_description="Provides current weather conditions, forecasts, and climate information for specific locations."
)

Calculator = Agent(
    name="Calculator",
    instructions="You are a powerful math assistant. Use the calculate_square tool to compute the square of the number provided in the user's query.",
    model=LitellmModel(model=MODEL, api_key=GEMINI_API_KEY),
    tools=[calculate_square],
    handoff_description="Performs mathematical calculations, specifically computing the square of numbers."
)

ManagerAgent = Agent(
    name="Manager",
    instructions="You are a manager agent. Analyze the user's query and decide whether to handle it yourself or delegate to a specialized agent. If the query is about weather, use the transfer_to_WeatherAssistant tool. If it involves calculating the square of a number, use the transfer_to_Calculator tool. For other general questions, you can answer directly.",
    model=LitellmModel(model=MODEL, api_key=GEMINI_API_KEY),
    handoffs=[WeatherAssistant, Calculator],
)

result = await Runner.run(
    ManagerAgent,
    "what is the weather of Tokyo"
)

print(result.final_output)




 Getting weather for Tokyo 


The weather in Tokyo is sunny and warm.



In [7]:
dir(result)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__dataclass_fields__',
 '__dataclass_params__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__match_args__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_last_agent',
 'context_wrapper',
 'final_output',
 'final_output_as',
 'input',
 'input_guardrail_results',
 'last_agent',
 'last_response_id',
 'new_items',
 'output_guardrail_results',
 'raw_responses',
 'to_input_list']

In [8]:
result.last_agent

Agent(name='Calculator', instructions="You are a powerful math assistant. Use the calculate_square tool to compute the square of the number provided in the user's query.", handoff_description='Performs mathematical calculations, specifically computing the square of numbers.', handoffs=[], model=<agents.extensions.models.litellm_model.LitellmModel object at 0x78130ce6b810>, model_settings=ModelSettings(temperature=None, top_p=None, frequency_penalty=None, presence_penalty=None, tool_choice=None, parallel_tool_calls=None, truncation=None, max_tokens=None, reasoning=None, metadata=None, store=None, include_usage=None, extra_query=None, extra_body=None, extra_headers=None), tools=[FunctionTool(name='calculate_square', description='', params_json_schema={'properties': {'val': {'title': 'Val', 'type': 'integer'}}, 'required': ['val'], 'title': 'calculate_square_args', 'type': 'object'}, on_invoke_tool=<function function_tool.<locals>._create_function_tool.<locals>._on_invoke_tool at 0x781313

In [9]:
result.last_agent.name

'Calculator'

## **Verify which work was done by which agent:**

In [10]:
result = await Runner.run(
    ManagerAgent,
    "what is the square of 26"
)

print(result.last_agent.name)
print(result.final_output)


print("__" * 30)

result = await Runner.run(
    ManagerAgent,
    "what is the weather of Gujar Khan"
)

print(result.last_agent.name)
print(result.final_output)


print("__" * 30)

result = await Runner.run(
    ManagerAgent,
    "who is thw founder of Pakistan"
)

print(result.last_agent.name)
print(result.final_output)



 Calculating square 


Calculator
The square of 26 is 676.

____________________________________________________________

 Getting weather for Gujar Khan 


WeatherAssistant
The weather in Gujar Khan is sunny and warm.

____________________________________________________________
Manager
The founder of Pakistan is Muhammad Ali Jinnah.



## **Handoff vs Handoffs:**

- **Handoff**: Jab aap ek agent se kaam doosre **ek specific agent** ko transfer karte hain, to yeh **handoff** hota hai. Yani, ek single action jisme kaam ek agent se doosre tak jata hai.

- **Handoffs**: Yeh tab use hota hai jab aap ek agent ke paas **multiple agents** ke options hote hain jinko woh kaam transfer kar sakta hai. Yani, yeh ek **list** ya collection hai jisme saare possible agents ke naam hote hain jinko handoff kiya ja sakta hai.

### **Aasan Alfaaz Mein**
- Agar aapko **sirf ek agent** ko kaam dena hai, to aap **handoff** ka use karenge. Maslan, "Is kaam ko Refund Agent ko do."
- Agar aapke agent ke paas **multiple agents** ke options hain aur woh decide karega ke kaam kisko dena hai, to yeh **handoffs** ke concept mein aata hai. Maslan, "Mere paas Refund Agent, Spanish Agent, aur FAQ Agent ke options hain, inme se kisi ko kaam do."

### **Chhoti Si Baat**
Documentation ke hisaab se, **handoffs** ek agent ka property hai jo batata hai ke uske paas kaunse agents available hain kaam transfer karne ke liye. Lekin jab actual kaam transfer hota hai, to woh ek **handoff** action hota hai, chahe ek agent ho ya multiple options mein se ek chuna jaye.

Toh aapka point sahi hai: **ek agent ke liye handoff, multiple ke liye handoffs ka concept**. Bas yeh yaad rakhein ke **handoffs** zyada tar options ki list ko represent karta hai, jabke **handoff** action ko.

-----------------
-----------------
-----------------

## **Handoff_description:**

Aasaan alfaaz mein, main aapko batata hoon ke **handoff_description** kya hai aur yeh kya kaam karta hai:

### **Handoff_description Kya Hai?**
- **Handoff_description** ek property ya field hai jo ek agent ke configuration mein hoti hai jab aap OpenAI Agents SDK use kar rahe hote hain.
- Yeh ek **string** hoti hai jo batati hai ke yeh agent kya kaam karta hai ya kis tarah ke tasks ke liye suitable hai.

### **Yeh Kya Kaam Karta Hai?**
- **Handoff_description** ka maqsad hai ek agent ke role ya purpose ko clearly define karna taake doosre agents (jaise triage agent) samajh sakein ke is agent ko kaunsa kaam dena chahiye.
- Jab ek agent (maslan, triage agent) decide karta hai ke kaam kisko handoff karna hai, to woh **handoff_description** ko padhta hai taake yeh pata chal sake ke kaunsa agent us kaam ke liye best hai.
- Yeh ek tarah ka label ya summary hota hai jo agent ke kaam ko describe karta hai.

### **Example Se Samjhein**
Maano aapke paas teen agents hain:
1. **Refund Agent**: Iske `handoff_description` mein likha hai: `"Handles refund requests and processes returns."`
2. **Spanish Agent**: Iske `handoff_description` mein likha hai: `"Assists users with queries in Spanish."`
3. **FAQ Agent**: Iske `handoff_description` mein likha hai: `"Answers common questions about products."`

Ab agar ek user ka sawal aata hai ke "Mujhe refund chahiye," to triage agent in sab agents ke **handoff_description** ko dekhega aur samajh jayega ke yeh kaam Refund Agent ke liye hai. Is tarah woh kaam Refund Agent ko handoff kar dega.

### **Kaise Use Hota Hai?**
Link ke mutabiq, jab aap ek agent banate hain, to aap uska `handoff_description` set karte hain:
```python
refund_agent = Agent(
    name="Refund Agent",
    handoff_description="Handles refund requests and processes returns."
)
```

Yeh description triage agent ke liye ek hint ki tarah kaam karta hai jab woh decide karta hai ke kaam kisko bhejna hai.

### **Simple Summary**
- **Handoff_description** ek chhota sa explanation hota hai jo batata hai ke ek agent kya karta hai.
- Iska kaam hai triage agent ko yeh samjhana ke kaunsa agent kis kaam ke liye perfect hai, taake kaam sahi jagah handoff ho sake.

Bas itna hi! Yeh ek tarah ka guide ya tag hai jo handoff process ko asaan banata hai.

## **Q:Mujhe yeh samajhna hai ke ManagerAgent ek hi prompt mein do alag tasks—like "2 ka square kya hai aur Karachi mein weather kya hai?"—ko pehchan kar unhein alag-alag agents (Calculator aur WeatherAssistant) ko delegate kar sakta hai ya nahi?**

## **Answer:** Agar aap chahte hain ke ManagerAgent dono tasks handle kare, to yeh tareeqa hain:
 - ## 1. **Agents ko as a Tools Dena Manager Agent ko:** ManagerAgent ko khud __calculate_square__ aur __weather__ tools do, taake woh bina delegate kiye dono jawab de sake. Maslan:
    - ##  "2 ka square" → 4
    - ## "Karachi mein weather" → (weather data)
   ## Phir ManagerAgent khud dono results combine kar ke jawab dega.

In [12]:
@function_tool()
def weather(city: str):
    print(f"\n Getting weather for {city} \n\n")
    return f"The weather in {city} is sunny and warm"

@function_tool()
def calculate_square(val: int):
    print("\n Calculating square \n\n")
    return f"The square of {val} is {val * val}"


WeatherAssistant = Agent(
    name="WeatherAssistant",
    instructions="You are a powerful weather assistant. Use the weather tool to provide accurate weather information for the location specified in the user's query.",
    model=LitellmModel(model=MODEL, api_key=GEMINI_API_KEY),
    tools=[weather],
    handoff_description="Provides current weather conditions, forecasts, and climate information for specific locations."
)

Calculator = Agent(
    name="Calculator",
    instructions="You are a powerful math assistant. Use the calculate_square tool to compute the square of the number provided in the user's query.",
    model=LitellmModel(model=MODEL, api_key=GEMINI_API_KEY),
    tools=[calculate_square],
    handoff_description="Performs mathematical calculations, specifically computing the square of numbers."
)

ManagerAgent = Agent(
    name="Manager",
    instructions="You are a manager agent. Analyze the user's query and decide whether to handle it yourself or delegate to a specialized agent. If the query is about weather, use the transfer_to_WeatherAssistant tool. If it involves calculating the square of a number, use the transfer_to_Calculator tool. For other general questions, you can answer directly.",
    model=LitellmModel(model=MODEL, api_key=GEMINI_API_KEY),
    # handoffs=[WeatherAssistant, Calculator],
    tools=[
        WeatherAssistant.as_tool(tool_name="WeatherAssistant", tool_description="Provides current weather conditions, forecasts, and climate information for specific locations."),
        Calculator.as_tool(tool_name="Calculator", tool_description="Performs mathematical calculations, specifically computing the square of numbers.")
    ]
)

result = await Runner.run(
    ManagerAgent,
    "what is the square of 24 and what is te weather in Gujar Khan"
)

print(result.final_output)



 Calculating square 



 Getting weather for Gujar Khan 


The square of 24 is 576. The weather in Gujar Khan is sunny and warm.



Aap ne mujhe ek code diya hai aur poocha hai ke kya `ManagerAgent` dono jawab (square of 24 aur weather in Gujar Khan) ek saath de sakta hai. Maine code ko dekha aur run kiya, aur output bhi check kiya. Ab main aap ko batata hoon ke yeh kaise kaam karta hai.

Aap ke code mein teen agents hain:
- **`WeatherAssistant`**: Yeh ek agent hai jo `weather` tool ke zariye kisi city ka weather batata hai. Abhi ke liye, yeh hamesha "sunny and warm" return karta hai.
- **`Calculator`**: Yeh ek agent hai jo `calculate_square` tool se kisi number ka square calculate karta hai.
- **`ManagerAgent`**: Yeh main agent hai jo user ke query ko analyze karta hai aur decide karta hai ke kaunsa tool use karna hai.

`ManagerAgent` ke paas do tools hain:
1. `WeatherAssistant` (jo as tool banaya gaya hai)
2. `Calculator` (jo bhi as tool banaya gaya hai)

Jab aap ka query "what is the square of 24 and what is te weather in Gujar Khan" `ManagerAgent` ko diya jata hai, toh yeh query mein do hisson ko pehchanta hai:
- "Square of 24" → Yeh calculation se related hai.
- "Weather in Gujar Khan" → Yeh weather se related hai.

`ManagerAgent` ke instructions ke mutabiq, woh query ko analyze karta hai aur phir dono tools ko call karta hai:
- `Calculator` tool ko 24 ke square ke liye (jo 576 return karta hai).
- `WeatherAssistant` tool ko Gujar Khan ke weather ke liye (jo "sunny and warm" return karta hai).

Phir `ManagerAgent` dono results ko combine karke ek jawab deta hai:  
**"The square of 24 is 576 and the weather in Gujar Khan is sunny and warm."**

### Yeh Kaise Kaam Karta Hai?
- `Agent.as_tool()` method se `WeatherAssistant` aur `Calculator` ko tools mein convert kiya gaya hai.
- `ManagerAgent` in tools ko directly call kar sakta hai, bina kisi handoff ke (handoff ka matlab hota hai ke control doosre agent ko de dena, lekin yahan aisa nahi ho raha).
- Underlying language model (LLM) query ko samajhta hai aur decide karta hai ke kaunse tools call karne hain. Chunki query mein dono tasks hain, toh dono tools call hote hain.

### Aap Ka Output
Aap ne jo output diya—"The square of 24 is 576 and the weather in Gujar Khan is sunny and warm"—bilkul sahi hai. Yeh dikhata hai ke `ManagerAgent` ne dono kaam ek saath handle kar liye.



# __Re-run the Loop (most important):__

Aap nay mujhse poocha ke "re-run the loop" ka kya matlab hai OpenAI Agents SDK ke context mein, aur aap beginner level par samajhna chahte hain. Main aap ko yeh concept bilkul aasaan zubaan mein samjhaata hoon.

### "Re-run the Loop" Kya Hai?
OpenAI Agents SDK mein, jab aap ek agent ko chalate hain (jaise `Runner.run()` method ke zariye), toh yeh ek loop ya chakkar mein kaam karta hai. Yeh loop tab tak chalta rehta hai jab tak agent ko final jawab na mil jaye. Har baar jab loop chalti hai, agent kuch steps follow karta hai:

1. **AI Model Ko Call Karna**: Agent pehle Large Language Model (LLM) ko call karta hai, jo AI ka dimagh hota hai. Yeh LLM aap ke sawal ya input ke mutabiq decide karta hai ke aage kya karna hai.
2. **Tools Ya Handoffs Ka Istemaal**: LLM decide kar sakta hai ke koi tool use kare (jaise weather check karne ke liye) ya kisi doosre agent ko kaam de (handoff kare).
3. **Check Karna Ke Jawab Pura Hua Ya Nahi**: Har action ke baad, agent dekhta hai ke kya yeh final jawab hai. Agar hai, toh loop ruk jati hai. Agar nahi, toh loop dobara chalti hai—ise hi "re-run the loop" kehte hain.

**"Re-run the loop"** ka matlab hai ke agent ko apna process dobara shuru karna padta hai jab tak woh final jawab tak na pohonch jaye. Yeh tab hota hai jab:
- **Handoff Hota Hai**: Agar LLM kehta hai ke yeh kaam kisi doosre agent ke liye hai, toh woh handoff karta hai aur loop naye agent ke saath dobara chalti hai.
- **Tool Ka Result Aata Hai**: Agar LLM koi tool use karta hai (jaise weather tool), toh tool ka result aane ke baad loop dobara chalti hai taake LLM us result ko samajh sake.

Loop tab rukti hai jab LLM final jawab de deta hai bina kisi tool ya handoff ke.

### Aasaan Misal
Maan lijiye aap agent se poochte hain, "Tokyo mein weather kya hai?"

1. **Pehla Loop**:
   - Agent LLM ko call karta hai.
   - LLM kehta hai, "Mujhe weather tool use karna hai."
   - Tool kehta hai, "Tokyo mein sunny aur warm hai."
   - Abhi final jawab nahi hai, kyunki tool ka result process karna baki hai.

2. **Doosra Loop (Re-run)**:
   - Agent dobara LLM ko call karta hai, ab tool ke result ("sunny aur warm") ke saath.
   - LLM dekhta hai ke yeh final jawab hai.
   - Loop ruk jati hai aur aap ko jawab milta hai: "Tokyo mein sunny aur warm hai."

Is misal mein, loop do baar chali: pehli baar tool ko call karne ke liye, aur doosri baar result ko final jawab banane ke liye. Yeh doosri baar chalna hi "re-run the loop" hai.

### Beginner Ke Liye Samajh
Sochiye agent ek dost hai jo aap ka sawal solve kar raha hai. Woh har baar kuch karta hai—jaise kisi se poochta hai ya koi cheez check karta hai—phir dekhta hai ke kya jawab mil gaya. Agar jawab adhoora hai, toh woh naye information ke saath dobara try karta hai. Yeh dobara try karna hi "re-run the loop" hai.

Link ke mutabiq, documentation mein likha hai: "If the LLM does a handoff, we update the current agent and input, and re-run the loop." Yani, agar handoff hota hai, toh naye agent aur input ke saath loop dobara chalti hai.
